File: commands/commands.go
package commands

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/scode/saltybox/preader"
	"github.com/scode/saltybox/secretcrypt"
	"github.com/scode/saltybox/varmor"
)

func encryptBytes(passphrase string, plaintext []byte) (string, error) {
	cipherBytes, err := secretcrypt.Encrypt(passphrase, plaintext)
	if err != nil {
		return "", fmt.Errorf("encryption failed: %w", err)
	}

	varmoredString := varmor.Wrap(cipherBytes)

	return varmoredString, nil
}

func Encrypt(inpath string, outpath string, preader preader.PassphraseReader) error {
	plaintext, err := os.ReadFile(inpath)
	if err != nil {
		return fmt.Errorf("failed to read from %s: %w", inpath, err)
	}

	passphrase, err := preader.ReadPassphrase()
	if err != nil {
		return err
	}
	encryptedString, err := encryptBytes(passphrase, plaintext)
	if err != nil {
		return fmt.Errorf("encryption failed: %w", err)
	}

	err = os.WriteFile(outpath, []byte(encryptedString), 0600)
	if err != nil {
		return fmt.Errorf("failed to write to %s: %w", outpath, err)
	}

	return nil
}

func decryptString(passphrase string, encryptedString string) ([]byte, error) {
	cipherBytes, err := varmor.Unwrap(encryptedString)
	if err != nil {
		return nil, fmt.Errorf("failed to unarmor: %w", err)
	}

	plaintext, err := secretcrypt.Decrypt(passphrase, cipherBytes)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt: %w", err)
	}

	return plaintext, nil
}

func Decrypt(inpath string, outpath string, preader preader.PassphraseReader) error {
	varmoredBytes, err := os.ReadFile(inpath)
	if err != nil {
		return fmt.Errorf("failed to read from %s: %w", inpath, err)
	}

	passphrase, err := preader.ReadPassphrase()
	if err != nil {
		return err
	}
	plaintext, err := decryptString(passphrase, string(varmoredBytes))
	if err != nil {
		return fmt.Errorf("failed to decrypt: %w", err)
	}

	err = os.WriteFile(outpath, plaintext, 0600)
	if err != nil {
		return fmt.Errorf("failed to write to %s: %w", outpath, err)
	}

	return nil
}

func Update(plainfile string, cryptfile string, pr preader.PassphraseReader) (err error) {
	// Decrypt existing file in order to validate that the provided passphrase is correct,
	// in order to prevent accidental changing of the passphrase (but we discard the plain
	// text).
	varmoredBytes, err := os.ReadFile(cryptfile)
	if err != nil {
		return fmt.Errorf("failed to read from %s: %w", cryptfile, err)
	}

	cachingPreader := preader.NewCaching(pr)

	passphrase, err := cachingPreader.ReadPassphrase()
	if err != nil {
		return err
	}
	_, err = decryptString(passphrase, string(varmoredBytes))
	if err != nil {
		return fmt.Errorf("failed to decrypt: %w", err)
	}

	// Encrypt contents into the target file using atomic semantics (write to tempfile, fsync()
	// and rename). This guarantees that the resulting file will either be the old file or the new
	// file, but never corrupt (assuming a correctly functioning filesystem I/O stack).
	cryptDir, _ := filepath.Split(cryptfile)

	tmpfile, err := os.CreateTemp(cryptDir, "saltybox-update-tmp")
	if err != nil {
		return fmt.Errorf("failed to create tempfile: %w", err)
	}
	defer func(fname string) {
		if _, localErr := os.Stat(fname); !os.IsNotExist(localErr) {
			err = os.Remove(fname)
		}
	}(tmpfile.Name())
	defer func(tmpfile *os.File) {
		err = tmpfile.Close()
	}(tmpfile)

	err = Encrypt(plainfile, tmpfile.Name(), cachingPreader)
	if err != nil {
		return fmt.Errorf("failed to encrypt: %w", err)
	}

	// Re-open the file to ensure that we are Sync():ing the correct file. Technically this is not
	// required because passphraseEncryptFile() will cause the target file to be truncated rather than recreated.
	// However, let's defensively avoid relying on that subtle behavior and re-open the file.
	reopenedTmpFile, err := os.Open(tmpfile.Name())
	if err != nil {
		return fmt.Errorf("failed to re-open tempfile after encryption: %w", err)
	}
	defer func(f *os.File) {
		err = f.Close()
	}(reopenedTmpFile)

	err = reopenedTmpFile.Sync()
	if err != nil {
		return fmt.Errorf("failed to sync file prior to rename: %w", err)
	}

	err = os.Rename(reopenedTmpFile.Name(), cryptfile)
	if err != nil {
		return fmt.Errorf("failed to rename to target file: %w", err)
	}

	return nil
}

File: saltybox-rs/src/commands.rs
use std::fs::{self, File, OpenOptions};
use std::io::Write;
use std::path::Path;

use crate::preader::PassphraseReader;
use crate::secretcrypt;
use crate::varmor;

fn encrypt_bytes(passphrase: &str, plaintext: &[u8]) -> anyhow::Result<String> {
    let cipher_bytes = secretcrypt::encrypt(passphrase, plaintext)?;
    let varmored = varmor::wrap(&cipher_bytes);
    Ok(varmored)
}

pub fn encrypt(inpath: &Path, outpath: &Path, mut pr: impl PassphraseReader) -> anyhow::Result<()> {
    let plaintext = fs::read(inpath)
        .map_err(|e| anyhow::anyhow!("failed to read from {}: {}", inpath.display(), e))?;
    let passphrase = pr.read_passphrase()?;
    let encrypted = encrypt_bytes(&passphrase, &plaintext)?;

    // Write with 0600 permissions best-effort
    let mut f = OpenOptions::new().create(true).truncate(true).write(true).open(outpath)
        .map_err(|e| anyhow::anyhow!("failed to write to {}: {}", outpath.display(), e))?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = fs::set_permissions(outpath, fs::Permissions::from_mode(0o600));
    }
    f.write_all(encrypted.as_bytes())
        .map_err(|e| anyhow::anyhow!("failed to write to {}: {}", outpath.display(), e))?;
    Ok(())
}

fn decrypt_string(passphrase: &str, encrypted_string: &str) -> anyhow::Result<Vec<u8>> {
    let cipher_bytes = varmor::unwrap(encrypted_string)
        .map_err(|e| anyhow::anyhow!("failed to unarmor: {}", e))?;
    let plaintext = secretcrypt::decrypt(passphrase, &cipher_bytes)
        .map_err(|e| anyhow::anyhow!("failed to decrypt: {}", e))?;
    Ok(plaintext)
}

pub fn decrypt(inpath: &Path, outpath: &Path, mut pr: impl PassphraseReader) -> anyhow::Result<()> {
    let varmored_bytes = fs::read(inpath)
        .map_err(|e| anyhow::anyhow!("failed to read from {}: {}", inpath.display(), e))?;
    let passphrase = pr.read_passphrase()?;
    let plaintext = decrypt_string(&passphrase, &String::from_utf8_lossy(&varmored_bytes))?;
    let mut f = OpenOptions::new().create(true).truncate(true).write(true).open(outpath)
        .map_err(|e| anyhow::anyhow!("failed to write to {}: {}", outpath.display(), e))?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        let _ = fs::set_permissions(outpath, fs::Permissions::from_mode(0o600));
    }
    f.write_all(&plaintext)
        .map_err(|e| anyhow::anyhow!("failed to write to {}: {}", outpath.display(), e))?;
    Ok(())
}

pub fn update(plainfile: &Path, cryptfile: &Path, pr: impl PassphraseReader) -> anyhow::Result<()> {
    // Validate passphrase by decrypting existing file first
    let varmored_bytes = fs::read(cryptfile)
        .map_err(|e| anyhow::anyhow!("failed to read from {}: {}", cryptfile.display(), e))?;
    let mut caching = crate::preader::CachingPassphraseReader::new(pr);
    let passphrase = caching.read_passphrase()?;
    let _ = decrypt_string(&passphrase, &String::from_utf8_lossy(&varmored_bytes))
        .map_err(|e| anyhow::anyhow!("failed to decrypt: {}", e))?;

    // Atomic replace via tempfile, fsync and rename
    let crypt_dir = cryptfile.parent().unwrap_or_else(|| Path::new("."));
    let tmp = tempfile::Builder::new()
        .prefix("saltybox-update-tmp")
        .tempfile_in(crypt_dir)?;

    // Write encrypted contents to tmp path using Encrypt
    let tmp_path = tmp.path().to_path_buf();
    // Ensure file exists with correct perms before writing
    {
        let _f = File::create(&tmp_path)?;
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let _ = fs::set_permissions(&tmp_path, fs::Permissions::from_mode(0o600));
        }
    }

    encrypt(plainfile, &tmp_path, caching)?;

    // Re-open to ensure we fsync the correct file
    let reopened = OpenOptions::new().read(true).write(true).open(&tmp_path)?;
    reopened.sync_all()?;

    // Persist moves it into place atomically
    tmp.persist(cryptfile)?;
    Ok(())
}

